Create UI components and stories for desktop SKU

zenparsing opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/8738

This PR adds React components required by the SKU feature on desktop, and supporting storybook stories. String localization and WebUI will be added in a follow-up PRs.

For now, all components added in this PR are located under `ui/components/checkout` in order to isolate them from existing shared components. In the future we can move components up a folder if they become shared.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

```
npm run storybook
```

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/formSection/style.ts:
```diff
@@ -0,0 +1,24 @@
+export const Box = styled.div`
+  border-radius: 8px;
+  background: #F7F9FB;
+  box-shadow: 1px -1px 6px 0 rgba(12, 23, 33, 0.20);
```
I think this should be `0 0 6px 0 rgba(12, 23, 33, 0.20);`. Even though that sketch has this values in the browser it doesn't look the same.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/index.tsx:
```diff
@@ -0,0 +1,86 @@
+  return (
+    <AddFundsButtonContainer>
+      {/* Plus icon */}
+      <svg width='12' height='12' xmlns='http://www.w3.org/2000/svg'>
```
let's put this icon in brave-ui repo, so that we can use it directly

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/index.tsx:
```diff
@@ -0,0 +1,86 @@
+      <svg width='12' height='12' xmlns='http://www.w3.org/2000/svg'>
+        <path d='M6 0a.75.75 0 01.75.75v4.5h4.5a.75.75 0 110 1.5h-4.5v4.5a.75.75 0 11-1.5 0v-4.5H.75a.75.75 0 110-1.5h4.5V.75A.75.75 0 016 0z' fill='#FFF' fill-rule='evenodd' />
+      </svg>
+      <Button text={'Add Funds'} size={'medium'} onClick={props.onClick} />
```
button has property `icon` so you can just use that

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,132 @@
+
+export function AddFundsPanel (props: AddFundsPanelProps) {
+  if (props.amountOptions.length === 0) {
+    throw new Error('props.amountOptions cannot be an empty array')
```
I think we should not throw as it will display blank component. I think we should handle this more softly

zenparsing replied:
Done - if given an empty array, the component will simply not display any option buttons.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,132 @@
+      <CurrentBalance>
+        <div>
+          Current balance
+          <span className='balance-bat'>{props.walletBalance.toFixed(1)} BAT</span>
```
wallet balance should already be formatted  when it comes in

zenparsing replied:
Done. I've kept the `walletBalance` as a number in `PaymentMethodPanel` though, as it compares that value with the total order amount.

NejcZdovc commented:
Please use theme color, no custom colors

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,138 @@
+
+  button {
+    float: right;
+    width: 231px;
```
we shouldn't set width as if long translation comes in it will go over. Outer should set button width or set min-width to this one

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,138 @@
+  button {
+    float: right;
+    width: 231px;
+    background: ${p => p.theme.color.brandBat};
```
you don't need any of this as button already rewards button https://brave.github.io/brave-ui/?path=/story/components--button

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,138 @@
+export const ExchangeRateDisplay = styled.div`
+  font-size: 12px;
+  color: #8B97A9;
+  position: absolute;
```
let's avoid absolute as with long translations we could have problems, maybe just flex?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/icons.tsx:
```diff
@@ -0,0 +1,13 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
please move this in brave-ui repo

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/index.tsx:
```diff
@@ -0,0 +1,85 @@
+      <CardNumber>
+        <label>
+          Card number
+          <input ref={cardNumberRef} autoComplete='cc-number' />
```
can we use input from brave-ui as it already has handler for icons?

zenparsing replied:
Unfortunately, `<Input />` doesn't currently expose a `ref` prop and it doesn't accept things like `autoComplete`. https://github.com/brave/brave-ui/issues/373

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/assets/dialog_bg.svg:
```diff
@@ -0,0 +1,43 @@
+<svg width="194" height="228" xmlns="http://www.w3.org/2000/svg">
```
add licence

NejcZdovc commented:
Can we avoid using things like `first-child` and `last-child`. Problem is that things can go wrong really fast when we need to add some items in between or move things around. Would really prefer styled div so that is clearly connected to css

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/termsOfSale/index.tsx:
```diff
@@ -0,0 +1,26 @@
+    after = ''
+  ] = props.text.split(/\$\d+/g)
+
+  // TODO: link URL
```
TODO

zenparsing replied:
Still waiting for a URL for this link.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/termsOfSale/index.tsx:
```diff
@@ -0,0 +1,26 @@
+  text: string
+}
+
+export function TermsOfSale (props: TermsOfSaleProps) {
```
This component can be removed and just put the whole text in translations. Example of translation https://github.com/brave/brave-core/blob/a306c7e071f6f3935c32c5e88a8863d0d19badf7/components/resources/brave_components_strings.grd#L308. We do it like this so other languages can move link around as order of words is not the same in all languages.

zenparsing commented:
@NejcZdovc CSS is looking much cleaner now - thanks!

Remaining tasks:

- Add icons to `brave-ui`: branch is ready, waiting for push access to repository
- Add URL for "Terms of Sale": waiting on final URL.
- Fix `Input` in `brave-ui` to forward refs and `autoComplete` props.

Otherwise, I think the code is ready for another look.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/style.ts:
```diff
@@ -41,12 +43,12 @@ export const ActionPanel = styled.div`
 `
 
-export const BatAmount = styled.span`
+export const BatAmount = styled.span<{ locale: LocaleData }>`
```
can we please move this in tsx, so that css is pure

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+        <CreditCardForm handleRef={creditCardFormRef} />
+      </FormSection>
+      <PurchaseButtonRow>
+        <div>
```
is div needed?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+        <div>
+          <GoBackLink onClick={props.onCancel} />
+        </div>
+        <div>
```
is div needed?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+        title={
+          <>
+            <div>1. ${locale.get('selectAmountToAdd')}</div>
+            <div>
```
is div needed?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+        {locale.get('addFundsSubtitle')}
+      </Subtitle>
+      <CurrentBalance>
+        <div>
```
is div needed?

zenparsing replied:
Yes - it's a flex child.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+      <FormSection
+        title={
+          <>
+            <div>1. ${locale.get('selectAmountToAdd')}</div>
```
remove `$`

NejcZdovc replied:
`1.` should be part of translations as numbers can be translated as well

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+          setSelectedAmount={setSelectedAmount}
+        />
+      </FormSection>
+      <FormSection title={`2. ${locale.get('enterCreditCardInfo')}`}>
```
`2` should be part of translation as well

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/index.tsx:
```diff
@@ -0,0 +1,175 @@
+          <GoBackLink onClick={props.onCancel} />
+        </div>
+        <div>
+          <Button
```
I see that we have design for `PurchaseButtonRow button`. Let's try to use approach if possible. Name this one something like `AddFundsButton`. Then in style file you can do this

```
 export const PurchaseButtonRow = styled(Button as ComponentType<ButtonProps>)`
    padding-left: 42px;
    padding-right: 42px;
`

```

This way we remove nesting in `style.ts`

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,130 @@
+  text-align: right;
+  margin-top: -2px;
+
+  svg {
```
see example for button, you can do the same thing with an icon. 

```
export const BatIcon = styled(BatColorIcon)`
  height: 18px;
  width: 18px;
  vertical-align: text-bottom;
`
```

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,130 @@
+  justify-content: center;
+  flex-wrap: wrap;
+
+  > div {
```
this css can be moved to `AmountOptionContainer`

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,130 @@
+`
+
+export const AmountOptionContainer = styled.div<{ selected?: boolean }>`
+  button {
```
same as above, let's use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/style.ts:
```diff
@@ -0,0 +1,77 @@
+  display: grid;
+  grid-template-columns: 1fr 1fr;
+
+  input {
```
move this in separate styled component

zenparsing replied:
Unfortunately, the version of styled-components that we're on doesn't support `ref` the way that I need. We can probably take care of this when we fix the other `input` issues for this component.

NejcZdovc replied:
ok 👍

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/style.ts:
```diff
@@ -0,0 +1,77 @@
+  left: 0;
+  height: 0;
+
+  svg {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/style.ts:
```diff
@@ -0,0 +1,77 @@
+  grid-column: 1 / 3;
+  margin-top: -6px;
+
+  > label {
```
please use styled component

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/assets/dialog_bg.svg:
```diff
@@ -0,0 +1,47 @@
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
```
Please use same format of licensing that we have in other files, with `/*`

zenparsing replied:
Since this is SVG we'll need `<!--` for comments. Do we want a C-style comment inside of an XML comment?

NejcZdovc replied:
oh totally missed that is svg

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+    background-position: bottom right;
+  `}
+
+  a {
```
this can lead in overrides, do we need it here?

zenparsing replied:
I think it's easier to set a default link style here than to create inline styles for every link.

NejcZdovc replied:
ok

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+`
+
+export const DialogTitle = styled.span`
+  svg {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+
+export const TopBar = styled.div`
+  padding: 15px 22px 15px 35px;
+  position: absolute;
```
can we avoid absolute?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/creditCardForm/style.ts:
```diff
@@ -0,0 +1,77 @@
+  height: 0;
+
+  svg {
+    position: absolute;
```
can we avoid absolute?

zenparsing replied:
Going to try to remove most of these absolutes today - will need to re-check against designs as I go.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+
+export const DialogTitle = styled.span`
+  svg {
+    position: absolute;
```
can we avoid absolute?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+export const CloseButton = styled.button`
+  top: 22px;
+  right: 22px;
+  position: absolute;
```
can we avoid absolute?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/goBackLink/style.ts:
```diff
@@ -0,0 +1,21 @@
+  padding-left: 35px;
+
+  svg {
+    position: absolute;
```
can we avoid absolute?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/style.ts:
```diff
@@ -0,0 +1,85 @@
+  svg {
+    height: 22px;
+    width: 22px;
+    position: absolute;
```
can we avoid absolute?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/dialogFrame/style.ts:
```diff
@@ -0,0 +1,83 @@
+export const Content = styled.div`
+  padding: 51px 49px 34px;
+
+  h1 {
```
can this be moved into children? We can have h1 component that we can reuse in child components

zenparsing replied:
This is intentional. Are we not allowed to use regular HTML elements and cascade? It seems like more work and code to create styled wrappers with inline styles for every element.

zenparsing replied:
I went ahead and made the suggested change - thanks!

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/formSection/style.ts:
```diff
@@ -0,0 +1,26 @@
+  top: 0;
+  left: 0;
+
+  h2 {
```
please use styled component

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/localeContext/index.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const LocaleContext = React.createContext<LocaleData>({
+  get (key: string) {
+    return '?'
```
what does this `?` represent? If this is default value if translation is missing, we should be more exact like `Translation is missing` or something like this. `?` can be easily missed as expected result in UI

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/orderSummary/style.ts:
```diff
@@ -0,0 +1,54 @@
+  border-top: solid 1px ${p => p.theme.color.separatorLine};
+  border-bottom: solid 1px ${p => p.theme.color.separatorLine};
+
+  table {
```
please use styled components for all this elements

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/paymentComplete/assets/bat_refer.svg:
```diff
@@ -0,0 +1,36 @@
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
```
please use same license as other files, with `/*`

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/paymentComplete/style.ts:
```diff
@@ -0,0 +1,33 @@
+export const TextContainer = styled.div`
+  padding: 18px 21px 0 0;
+
+  h2 {
```
please use styled components

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/paymentProcessing/style.ts:
```diff
@@ -0,0 +1,20 @@
+  font-family: ${p => p.theme.fontFamily.body};
+  font-size: 15px;
+
+  svg {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useCreditCardPanel/index.tsx:
```diff
@@ -0,0 +1,111 @@
+          <CreditCardForm handleRef={creditCardFormRef} />
+        </FormSection>
+        <ConfirmButtonRow showBackLink={props.continueWithCard}>
+          {props.continueWithCard && <div><GoBackLink onClick={onBack} /></div>}
```
is this div needed?

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useCreditCardPanel/style.ts:
```diff
@@ -0,0 +1,54 @@
+  text-align: right;
+  min-width: 215px;
+
+  svg {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useCreditCardPanel/style.ts:
```diff
@@ -0,0 +1,54 @@
+  align-items: center;
+  justify-content: ${p => p.showBackLink ? 'space-between' : 'center'};
+
+  button {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/style.ts:
```diff
@@ -0,0 +1,85 @@
+    opacity: ${p => p.hasSufficientFunds ? '1' : '.5'};
+  }
+
+  svg {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/useWalletPanel/style.ts:
```diff
@@ -0,0 +1,85 @@
+  padding-right: 4px;
+  max-width: 185px;
+
+  button {
```
please use extend

NejcZdovc commented on components/brave_rewards/resources/ui/stories/checkout.tsx:
```diff
@@ -0,0 +1,153 @@
+    updated: 'Updated'
+  },
+  get (key: string) {
+    return this.strings[key] || '\u2753'
```
maybe `Missing translation` instead of question mark?

zenparsing commented:
The latest commits include the following changes:

- Created styled components for all previously nested style rules.
- Removed absolute positioning for all elements.
- Converted the "payment complete" vector graphic into an inline SVG instead of a bundled asset.
- Added "hasSufficientFunds" prop to `PaymentMethodPanel` and removed numeric comparisons.
- Tweaked credit card form validation code.

NejcZdovc commented on components/brave_rewards/resources/ui/components/checkout/addFundsPanel/style.ts:
```diff
@@ -0,0 +1,130 @@
+  background: ${p => p.selected ? p.theme.color.brandBat : 'transparent'};
+  border-color: ${p => p.theme.color.brandBat};
+  margin-bottom: 7px;
+  color: ${p => p.selected ? p.theme.palette.white : p.theme.color.brandBat};
```
please add font-size 15px here
